### PR TITLE
Improve annotation support

### PR DESCRIPTION
**Pull request by soc**
- Drop Java requirement for declaring "real" annotations
- Read Java annotations from Java source files to set
  Retention information correctly
- Introduce {Runtime,Constant}Annotation

### CODE REVIEW

**Comment by soc** [comment]
@retronym Could you have a look? I'm experiencing issues with separate compilation. :-/

**Comment by soc** [comment]
Nevermind, was just a stupid mistake.

**Comment by soc** [comment]
Ok, one of the last remaining issues is that `annotations` doesn't seem to be populated with platform annotations (I checked that the annotations are in the class file):

```
import scala.reflect.runtime.universe._
import scala.reflect.runtime.{currentMirror => cm}

class Bean {
  @JavaAnnotationWithNestedEnum_1(JavaAnnotationWithNestedEnum_1.Value.VALUE)
  def value = 1
}

object Test extends App {
  println(cm.staticClass("Bean").isCaseClass)
  println(typeOf[Bean].decl(TermName("value")).annotations)  // <--- annotations is empty!
  println(classOf[Bean].getMethod("value").getAnnotations.toList)
}
```

@xeno-by Do you have any idea?

**Comment by xeno-by** [comment]
Works for me in 2.11.7. Could you put up a reproduction project somewhere?

**Comment by soc** [comment]
@xeno-by It's related to the changes I made, I think I missed a place were annotations are read from the class file at runtime, but can't find it. :-/

File: src/library/scala/annotation/ConstantAnnotation.scala [hunk 1]
```diff
@@ -8,12 +8,14 @@
 
 package scala.annotation
 
-/** A base class for classfile annotations. These are stored as
- *  [[http://docs.oracle.com/javase/7/docs/technotes/guides/language/annotations.html#_top Java annotations]]]
- *  in classfiles.
+/** A base class for constant annotations.
```
Comment (soc) [inline src/library/scala/annotation/ConstantAnnotation.scala:11]: Better something like "marker"...

File: src/library/scala/annotation/ConstantAnnotation.scala [hunk 2]
```diff
@@ -8,12 +8,14 @@
...
  *
- *  @author  Martin Odersky
- *  @version 1.1, 2/02/2007
- *  @since 2.4
+ *  Constant annotations enforce that their arguments are constants.
+ *
+ *  These are available to the Scala type checker, even across different compilation units.
```
Comment (soc) [inline src/library/scala/annotation/ConstantAnnotation.scala:15]: Probably remove?

**Comment by soc** [comment]
@retronym This is the issue I talked about on Gitter.

**Comment by soc** [comment]
###### Note to myself

Jason said on Gitter: 

> I would debug (or println debug…) at https://github.com/scala/scala/blob/v2.12.0-M3/src/reflect/scala/reflect/runtime/JavaMirrors.scala#L683

**Comment by soc** [comment]
No, that didn't seem to do the trick. I added some println debugging and that codepath is never hit.

**Comment by retronym** [comment]
Can you share a minimal failing test? I'm still missing some context for
this problem.

**Comment by soc** [comment]
From scala/test/files/run/t6548:

``` java
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target({ElementType.ANNOTATION_TYPE, ElementType.METHOD, ElementType.FIELD,
    ElementType.TYPE, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
public @interface JavaAnnotationWithNestedEnum_1
{
    public Value value() default Value.VALUE;

    public enum Value
    {
        VALUE;
    }
}
```

``` scala
import scala.reflect.runtime.universe._
import scala.reflect.runtime.{currentMirror => cm}

class Bean {
  @JavaAnnotationWithNestedEnum_1(JavaAnnotationWithNestedEnum_1.Value.VALUE)
  def value = 1
}

object Test extends App {
  println(cm.staticClass("Bean").isCaseClass)
  println(typeOf[Bean].decl(TermName("value")).annotations)  // <--- annotations is empty!
  println(classOf[Bean].getMethod("value").getAnnotations.toList)
}
```

Result:

```
false
List()
List(@JavaAnnotationWithNestedEnum_1(value=VALUE))
```

**Comment by soc** [comment]
(I added the last `classOf` line to show that the information in question is in the classfile.)

File: src/library/scala/annotation/meta/constructor.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+**    / __/ __// _ | / /  / _ |    (c) 2002-2013, LAMP/EPFL             **
+**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
+** /____/\___/_/ |_/____/_/ | |                                         **
+**                          |/                                          **
+\*                                                                      */
+package scala.annotation.meta
+
+/**
+ * An annotation marked with `@constructor` signals that an annotation can be
+ * placed at a constructor.
```
Comment (lrytz) [inline src/library/scala/annotation/meta/constructor.scala]: the other annotations in the `meta` package don't specify a recommended target, but instead define the default target if there is no meta-annotation on an annotation usage.

File: src/library/scala/annotation/meta/constructor.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Additionally, marking an annotation with `@constructor` moves an annotation
+ * usage from a class declaration to the class' primary constructor if the
+ * allowed targets of the annotation in question doesn't also include class
+ * declarations.
```
Comment (lrytz) [inline src/library/scala/annotation/meta/constructor.scala]: I think we don't have a notion of "allowed targets", right? Also, is the semantics of the `@constructor` meta-annotation actually implemented in this PR?

File: src/library/scala/annotation/meta/constructor.scala (hunk 3 above)
Comment (soc) [inline src/library/scala/annotation/meta/constructor.scala]: Mhh, looks I got some only slightly related changes in this PR. I think I'll remove those things for now, as the implementation is not ready.

Thanks for catching it!

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala [hunk 5]
```diff
@@ -321,35 +337,102 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
...
+      accept(EQUALS)
+      val elementValue = qualIdOrClassLiteral()
+      AssignOrNamedArg(elementName, elementValue)
+    }
+
+    def elementValue(): Tree = in.token match {
+      case AT =>                       annotation()
+      case LBRACE =>                   elementValueArray()
+      case IDENTIFIER =>               qualIdOrClassLiteral()
+      case _ if isLiteral(in.token) => literal()
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:394]: Echoing my earlier comment: The java spec allows any constant expression as annotation value, not just literals, see https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.1 / https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.28

See also https://issues.scala-lang.org/browse/SI-5333

Comment (soc) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:394]: Thanks for reminding me, I'm currently still stuck on the other issue, but I will work on that afterwards.

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 6]
```diff
@@ -1140,6 +1148,7 @@ trait Definitions extends api.StandardDefinitions {
     // Meta-annotations
     lazy val BeanGetterTargetClass      = requiredClass[meta.beanGetter]
     lazy val BeanSetterTargetClass      = requiredClass[meta.beanSetter]
+    lazy val ConstructorTargetClass     = requiredClass[meta.constructor]
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala]: should be `getClassIfDefined("scala.annotation.meta.constructor")` so as not to crash with older standard libraries.

**Comment by retronym** [comment]
@soc The pickler [only writes](https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala#L202-L203) annotations into the `ScalaSignature` that extends `annotation.StaticAnnotation`. After your change to introduce `PlatformAnnotation`, java annotations are no longer taken as extending `StaticAnnotation`. (`ClassfileAnnotation` used to extend `StaticAnnotation`, now it doesn't).

As such, no annotations are recorded in the `ScalaSignature` for the method `value`.

You could change `AnnotationInfo#isStatic` to include the newly introduced `PlatformAnnotation`, or maybe have `PlatformAnnotation` extends `StaticAnnotation`. I haven't reviewed your design of the new annotation hierarchy enough to be able to suggest which way is better,

**Comment by soc** [comment]
Wow thanks! I'll make a change and try it again! Thanks! :-)

**Comment by soc** [comment]
@retronym I could adapt the condition, or adapt the annotation hierarchy, but I'm wondering if that makes sense because from a general point of view, we would duplicate the annotation information across Java class files and Scala signatures. Do you think it is not a huge problem, or should I find a different solution?

**Comment by retronym** [comment]
Duplication is fine. Most of the scalasig duplicates (albeit in a more precise way) the sort of information in the platform class file format.

**Comment by retronym** [comment]
... And the Scala typechecker exclusively uses the pickled info. Changing this to use hybrid sources of info would be a big undertaking.

**Comment by soc** [comment]
Ok, progress!

Two remaining issues:

(1) Java reflection seems to encounter duplicated annotations when defining annotations in Scala and evaluating the code with a toolbox:

For instance in https://github.com/soc/scala/blob/SI-8928-better-annotations/test/files/run/reflection-scala-annotations.scala:

```
new sann(1, scala.collection.immutable.List.apply[Int](1, 2))
java.lang.annotation.AnnotationFormatError: Duplicate annotation for class: interface java.lang.annotation.Retention: @java.lang.annotation.Retention(value=CLASS)
    at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:125)
    at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
    at java.lang.Class.createAnnotationData(Class.java:3521)
    at java.lang.Class.annotationData(Class.java:3510)
    at java.lang.Class.getAnnotations(Class.java:3446)
    at scala.reflect.runtime.JavaMirrors$JavaMirror.scala$reflect$runtime$JavaMirrors$JavaMirror$$$anonfun$26(JavaMirrors.scala:597)
    at scala.Option.flatMap(Option.scala:171)
    at scala.reflect.runtime.JavaMirrors$JavaMirror.loadAnnotation$1(JavaMirrors.scala:596)
    at scala.reflect.runtime.JavaMirrors$JavaMirror.loadBytes$1(JavaMirrors.scala:606)
    at scala.reflect.runtime.JavaMirrors$JavaMirror.unpickleClass(JavaMirrors.scala:613)
    at scala.reflect.runtime.SymbolLoaders$TopClassCompleter.scala$reflect$runtime$SymbolLoaders$TopClassCompleter$$$anonfun$2(SymbolLoaders.scala:26)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    at scala.reflect.internal.SymbolTable.slowButSafeEnteringPhaseNotLaterThan(SymbolTable.scala:263)
    at scala.reflect.runtime.SymbolLoaders$TopClassCompleter.complete(SymbolLoaders.scala:23)
    at scala.reflect.runtime.SymbolLoaders$TopClassCompleter.load(SymbolLoaders.scala:31)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$class.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$class$$$anonfun$8(SynchronizedSymbols.scala:140)
    at scala.reflect.runtime.Gil$class.gilSynchronized(Gil.scala:19)
    at scala.reflect.runtime.JavaUniverse.gilSynchronized(JavaUniverse.scala:16)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$class.gilSynchronizedIfNotThreadsafe(SynchronizedSymbols.scala:123)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$8.gilSynchronizedIfNotThreadsafe(SynchronizedSymbols.scala:168)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$class.typeParams(SynchronizedSymbols.scala:133)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$8.typeParams(SynchronizedSymbols.scala:168)
    at scala.reflect.internal.Types$NoArgsTypeRef.typeParams(Types.scala:1914)
    at scala.reflect.internal.Definitions$DefinitionsClass.fullyInitializeType(Definitions.scala:209)
    at scala.reflect.internal.Types$Type.toString(Types.scala:933)
    at scala.reflect.internal.Printers$TreePrinter.printTree(Printers.scala:476)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32(Printers.scala:529)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32$adapted(Printers.scala:526)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:32)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$TreePrinter.print(Printers.scala:526)
    at scala.reflect.internal.Printers$TreePrinter.printTree(Printers.scala:430)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32(Printers.scala:529)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32$adapted(Printers.scala:526)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:32)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$TreePrinter.print(Printers.scala:526)
    at scala.reflect.internal.Printers$TreePrinter.printTree(Printers.scala:457)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32(Printers.scala:529)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32$adapted(Printers.scala:526)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:32)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$TreePrinter.print(Printers.scala:526)
    at scala.reflect.internal.Printers$TreePrinter.printTree(Printers.scala:439)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32(Printers.scala:529)
    at scala.reflect.internal.Printers$TreePrinter.scala$reflect$internal$Printers$TreePrinter$$$anonfun$32$adapted(Printers.scala:526)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:32)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$TreePrinter.print(Printers.scala:526)
    at scala.reflect.api.Printers$class.render(Printers.scala:185)
    at scala.reflect.api.Universe.render(Universe.scala:61)
    at scala.reflect.api.Printers$class.show(Printers.scala:201)
    at scala.reflect.api.Universe.show(Universe.scala:61)
    at scala.reflect.api.Printers$class.treeToString(Printers.scala:193)
    at scala.reflect.api.Universe.treeToString(Universe.scala:61)
    at scala.reflect.api.Trees$TreeApi$class.toString(Trees.scala:178)
    at scala.reflect.internal.Trees$TreeContextApiImpl.toString(Trees.scala:102)
    at java.lang.String.valueOf(String.java:2994)
    at java.io.PrintStream.println(PrintStream.java:821)
    at scala.Console$.println(Console.scala:148)
    at scala.Predef$.println(Predef.scala:309)
    at Test$.delayedEndpoint$Test$1(reflection-scala-annotations.scala:15)
    at Test$delayedInit$body.apply(reflection-scala-annotations.scala:13)
    at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
    at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
    at scala.App$class.scala$App$class$$$anonfun$1(App.scala:76)
    at scala.App$class.scala$App$class$$$anonfun$1$adapted(App.scala:76)
    at scala.collection.immutable.List.foreach(List.scala:376)
    at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
    at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:44)
    at scala.App$class.main(App.scala:76)
    at Test$.main(reflection-scala-annotations.scala:13)
    at Test.main(reflection-scala-annotations.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at scala.reflect.internal.util.ScalaClassLoader$class.scala$reflect$internal$util$ScalaClassLoader$class$$$anonfun$8(ScalaClassLoader.scala:98)
    at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:32)
    at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:129)
    at scala.reflect.internal.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:98)
    at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:129)
    at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:22)
    at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:39)
    at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:29)
    at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:39)
    at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:65)
    at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
    at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
    at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
    at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

(2) It seems like reification needs to be adapted to annotations.

For instance in https://github.com/soc/scala/blob/SI-8928-better-annotations/test/files/run/reify_classfileann_a.scala compilation fails with:

```
reify_classfileann_a.scala:6: error: type mismatch;
 found   : scala.reflect.ClassTag[String]
 required: String
class ann(bar: String = "bar", quux: Array[String] = Array()) extends annotation.ClassfileAnnotation
                                                          ^
one error found
```

**Comment by retronym** [comment]
The code you've added to `typedClassDef` will run both in the `typer` and `erasure` phase, which will add the `RetentionPolicy` annotation twice. You could check if the annotation was present before you tried to automatically add it, and/or disable this logic if `isPastTyper`.

**Comment by soc** [comment]
Amazing, that resolves (1)!

**Comment by soc** [comment]
Do you have an idea regarding (2)?

**Comment by retronym** [comment]
I haven't figured out what changed to require this patch, but the problem can be solved by changing `tree2ConstArg` to look at the `Array.apply[T](...](<classtag>)` tree holistically. It was mistakenly trying to interpret the classtag argument as an array element.

``` diff
⚡ git diff head
diff --git a/src/compiler/scala/tools/nsc/typechecker/Typers.scala b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
index 5a2b12d..b1aa287 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -3680,7 +3680,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper

         // use of Array.apply[T: ClassTag](xs: T*): Array[T]
         // and    Array.apply(x: Int, xs: Int*): Array[Int]       (and similar)
-        case Apply(fun, args) =>
+        case treeInfo.Applied(fun, targs, args :: _) =>
           val typedFun = typed(fun, mode.forFunMode)
           if (typedFun.symbol.owner == ArrayModule.moduleClass && typedFun.symbol.name == nme.apply)
             pt match {
```

The problem was unrelated to `reify`, BTW.

**Comment by soc** [comment]
Thanks!

**Comment by soc** [comment]
Mhhh

```
reify_classfileann_b.scala:6: error: classfile annotation arguments have to be supplied as named arguments
class ann(bar: String = "bar", quux: Array[String] = Array(), baz: arg = new arg(bao = "arg")) extends annotation.ClassfileAnnotation
                                                                                       ^
reify_classfileann_b.scala:6: error: annotation arg is missing argument bao
class ann(bar: String = "bar", quux: Array[String] = Array(), baz: arg = new arg(bao = "arg")) extends annotation.ClassfileAnnotation
                                                                         ^
reify_classfileann_b.scala:6: error: expected annotation of type <error>, found <error>
class ann(bar: String = "bar", quux: Array[String] = Array(), baz: arg = new arg(bao = "arg")) extends annotation.ClassfileAnnotation
                                                                             ^
```

it goes away if I remove `(bao = "arg")`.

Inspecting `names`, `argss` and `args` in https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3686 reveals that the name is not preserved in them:

```
names: Set(value baz, value bar, value quux)
argss: List(List(val bar: String = "bar", val quux: Array[String] = scala.Array.apply[String]()((ClassTag.apply[String](classOf[java.lang.String]): scala.reflect.ClassTag[String])), val baz: arg = new arg("arg")))
args: List(val bar: String = "bar", val quux: Array[String] = scala.Array.apply[String]()((ClassTag.apply[String](classOf[java.lang.String]): scala.reflect.ClassTag[String])), val baz: arg = new arg("arg"))
```

**Comment by soc** [comment]
Isn't the call to `hasValue` in https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3687 turning the traversal into O(n²) business?

**Comment by retronym** [comment]
Named arguments are eliminated when typechecking applications. After typechecking, you can find the name by looking the name corresponding parameter symbol of the method being called.

``` scala
scala> { def foo(a: Int, b: Int = 0) = 0; foo(b = 0, a = 0)} //print

{
  def foo(a: scala.Int, b: scala.Int = 0) = 0;
  {
    val x$1: Int = 0;
    val x$2: Int = 0;
    foo(x$2, x$1)
  }
} // : Int

scala> { def foo(a: Int, b: Int = 0) = 0; foo(a = 0, b = 0)} //print
   {
  def foo(a: scala.Int, b: scala.Int = 0) = 0;
  foo(0, 0)
} // : Int

```

**Comment by adriaanm** [comment]
I'm going to move this to M5. I'd like to cut M4 as soon as the new trait encoding is ready.

**Comment by lrytz** [comment]
closing for now as it's not been active in a while, please re-open when you're working on it.

**Pull request closed without merging**